### ISSUE TITLE
Provide interface for differentiating w.r.t. physics (differentiable physics)

### ISSUE THREAD

**Issue by peci1** [gazebosim/gz-sim#2936]
## Desired behavior
It would be very interesting if gz-physics and gz-sim provided an interface for retrieving gradients w.r.t. physics. This is also known as "differentiable simulation" or "differentiable physics".

## Implementation suggestion
There are several use-cases for gradients w.r.t. physics, e.g.:

### Model identification

Compute gradient w.r.t. IMU pose, camera pose, the inertial matrix or mass of a part.

### Identification of physical parameters

Compute gradient w.r.t. friction coefficients, damping, slip coefficients etc.

### Identification of unobserved terrain

(C) Karel Zimermann 

https://github.com/user-attachments/assets/72f9e48c-adbe-446f-b3b1-b70fb066cb4f

### PINNs (Physics-Informed Neural Networks)

Gradient w.r.t physics is a part of the loss function.

Demo of optimizing a ballistic curve: https://gist.github.com/peci1/720e7df143dd20cd25296127c89868cf .

### Neuro-Symbolic Models

Basically Deep NNs with diff physics as one of the network modules, e.g. https://arxiv.org/abs/2309.09007 or https://arxiv.org/abs/2404.01596 .

### (Indirectly) differentiate w.r.t. control to find optimal policies

Differentiability of control is not directly related, but if you can get a derivative that takes physics into account, it has much higher value.

### Problems

There are also lots of problems when differentiating naturally non-smooth functions like rigid body contacts. A good reading is at https://arxiv.org/abs/2109.05143 . It seems that analytical gradients are bad when contacts come into play. More suitable could be implicit gradients or stochastically computed ones.

Also, usual ML pipelines need to compute gradient and then somehow "apply" it to the current state to get a new state (or a batch of new states). Would Gazebo be able to do something like that?

### Known diff simulators

- https://nvidia.github.io/warp/
- https://genesis-world.readthedocs.io/en/latest/index.html
- https://github.com/google/brax
- https://nimblephysics.org/ (diff fork of DART)
- https://stack-of-tasks.github.io/pinocchio/
- lots of others

CC @scpeters
